src/supervisor.adoc

```diff
@@ -1217,30 +1217,60 @@ Two schemes to manage the A and D bits are permitted:
-later than the explicit memory access, or any subsequent explicit memory
-access to that virtual page by the local hart. The ordering on loads and
-stores provided by FENCE instructions and the acquire/release bits on
-atomic instructions also orders the PTE updates associated with those
-loads and stores as observed by remote harts.
-+
-The PTE update is not required to be atomic with respect to the explicit
-memory access that caused the update, and the sequence is interruptible.
-However, the hart must not perform the explicit memory access before the
-PTE update is globally visible.
+* When a virtual page is accessed and the A bit is clear, the PTE is
+  updated to set the A bit.  When the virtual page is written and the D
+  bit is clear, the PTE is updated to set the D bit. When G-stage address
+  translation is in use and is not Bare, the G-stage virtual pages may be
+  accessed or written by implicit accesses to VS-level memory management
+  data structures, such as page tables.                                 +
+                                                                        +
+  When two-stage address translation is in use, an explicit access may
+  cause both VS-stage and G-stage PTEs to be updated. The following rules
+  apply to all PTE updates caused by an explicit or an implicit memory
+  accesses.                                                             +
+                                                                        +
+  The PTE update must be atomic with respect to other accesses to the
+  PTE, and must atomically perform all tablewalk checks for that leaf
+  PTE as part of, and before, conditionally updating the PTE value.
+  Updates of the A bit may be performed as a result of speculation, even
+  if the associated memory access ultimately is not performed
+  architecturally. However, updates to the D bit, resulting from an
+  explicit store, must be exact (i.e., non-speculative), and observed in
+  program order by the local hart. When two-stage address translation is
+  active, updates of the D bit in G-stage PTEs may be performed as a
+  result of speculative updates of the A bit in VS-stage PTEs.          +
+                                                                        +
+  The PTE update must appear in the global memory order before the
+  memory access that caused the PTE update and before any subsequent
+  explicit memory access to that virtual page by the local hart. The
+  ordering on loads and stores provided by FENCE instructions and the
+  acquire/release bits on atomic instructions also orders the PTE updates
+  associated with those loads and stores as observed by remote harts.   +
+                                                                        +
+  The PTE update is not required to be atomic with respect to the memory
+  access that caused the update and a trap may occur between the PTE
+  update and the memory access that caused the PTE update. If a trap
```

Comment:
Would it be possible to extend this to a "trap or failing sc", so that if a sc causes a PTE update to set the D bit, and that page table update interferes with the sc such that the reservation is lost, the implementation does not need to paradoxically not have updated the page table?

Consider in particular the case where the physical address of the PTE and the physical address written by the sc are the same or in the same cache line, and the page table walker does not have a connection to the L1 data cache.

Reply:
The A/D updates occur as part of the address  translation process which is a precursor for the instruction to execute. This process does  not treat instructions that have conditional execution or conditional trapping behavior such as masked vector stores, store conditional, fault-only-first vector loads, or compare-and-swap differently for the purpose of setting A and/or D bit.  So a special mention for sc may not be required.

Reply:
The A extension spec says "If the SC.W fails, the instruction does not write to memory".  I am interpreting that to mean that it does not cause "the virtual page is written" for the purposes of this section.

Reply:
Setting D bit is a property of the type of access the instruction could make i.e. a store access and not how the instruction executes. The setting of the D bit is not conditional on the store instructions conditional behavior -  such as the cases mentioned  -  for instance a masked  vector store does not update the elements in memory.

Reply:
What in the diff supports the interpretation of setting D bits based on accesses an instruction "could" make?

For masked vector stores, especially in the large stride or indexed case, it's important to not require that an implementation perform page table walks for masked-off elements.  The V spec already requires that masked-off elements not generate exceptions, so it would be extremely odd to specify that setting the D bit occurs regardless of the mask value.  We should probably allow imprecise D bit setting in that case, since many implementations will want to perform page table checks for the unit or small stride case optimistically without depending on the mask value.

Reply:
This diff does not as this is detailed in the existing section 5.3.2.  The page table walk algorithm and the steps for when the D bit are set  does not include conditionally setting of A and/or D bit based on conditional behavior of instructions.

Reply:
5.3.2 does not mention instructions at all, except for SFENCE.VMA.

Reply:
Sorry if I was not being clear. I referred to 5.3.2 to say that the algorithm is agnostic of any conditional execution behavior of the instruction itself i.e. there is no dependence on any conditional execution behavior of instructions for the purpose of setting A and/or D bits

Reply:
Is an indexed vector store required to perform address translations independent of the mask values?

Is an indexed vector store required to perform address translations independent of the VL value?

Is an indexed vector store required to perform address translations independent of the LMUL value?

Reply:
Since A&D functionality is orthogonal to any specific load/store instructions, the answer should come from the vector spec.  Also note that the question of updating A&D bits only arises for accesses that pass all MMU/PMP/PMA permission and access restrictions.

So, in short (and assuming one is accessing idempotent memory), "masked" elements (whether due to vstart, mask, or vl) may be accessed.  What the spec generally constrains is "do not raise exceptions and do not update the destination vector register".  Hence masked elements are not required to be translated, but are allowed be translated and accessed - and A&D bits may then be set as a result .